src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.CurlResponseMessage.cs

```diff
@@ -0,0 +1,500 @@
+                        Debug.Assert(_remainingDataCount >= 0, "The remaining count should never go negative");
+                        Debug.Assert(_remainingDataOffset <= _remainingData.Length, "The remaining offset should never exceed the buffer size");
+
+                        VerboseTrace("Copied to task: " + bytesToCopy + " (Position: " + _position + ")");
+                        return Task.FromResult(bytesToCopy);
+                    }
+
+                    // If the stream has already been completed, complete the read immediately.
+                    if (_completed == s_completionSentinel)
+                    {
+                        VerboseTrace("Completed successfully after stream completion");
+                        return s_zeroTask;
+                    }
+
+                    // Finally, the stream is still alive, and we want to read some data, but there's no data 
+                    // in the buffer so we need to register ourself to get the next write.
+                    if (cancellationToken.CanBeCanceled)
+                    {
+                        // If the cancellation token is cancelable, then we need to register for cancellation.
+                        // We creat a special CancelableReadState that carries with it additional info:
+                        // the cancellation token and the registration with that token.  When cancellation
+                        // is requested, we schedule a work item that tries to remove the read state
+                        // from being pending, canceling it in the process.  This needs to happen under the
+                        // lock, which is why we schedule the operation to run asynchronously: if it ran
+                        // synchronously, it could deadlock due to code on another thread holding the lock
+                        // and calling Dispose on the registration concurrently with the call to Cancel
+                        // the cancellation token.  Dispose on the registration won't return until the action
+                        // associated with the registration has completed, but if that action is currently
+                        // executing and is blocked on the lock that's held while calling Dispose... deadlock.
+                        var crs = new CancelableReadState(buffer, offset, count, this, cancellationToken);
+                        crs._registration = cancellationToken.Register(s1 => Task.Factory.StartNew(s2 =>
+                        {
+                            var crsRef = (CancelableReadState)s2;
+                            lock (crsRef._stream._lockObject)
+                            {
+                                if (crsRef._stream._pendingReadRequest == crsRef)
+                                {
+                                    crsRef.TrySetCanceled(crsRef._token);
+                                    crsRef._stream._pendingReadRequest = null;
```

Comment:
It doesn't actually matter in this case.  ClearPendingReadRequest does two things: Disposes of the CTR and nulls out the field.  Since we're in the callback for cancellation being requested, we know that cancellation was requested, and thus Dispose'ing of the CTR will be a nop.  But for consistency and to avoid confusion, I changed it to use ClearPendingReadRequest, as while Dispose'ing the CTR won't help, it also won't hurt.